### PR TITLE
Add mypy type hinting check

### LINKED ISSUE #1002: Type hint linting
It should be possibly to lint the pyspec type hints. See [here](https://github.com/ethereum/eth2.0-specs/pull/996#issuecomment-487065832):

> You can run mypy/a linter on the output pyspec file

### PR DESCRIPTION
**Pull request by hwwhww**
1. Fix #1002
2. Refactor hash cache
3. Note: since mypy doesn't support `X = NewType('X', NONE-BASE-TYPE)` (https://github.com/python/mypy/issues/4615), I changed `uint64` to `class` to make the check pass. I'll benchmark it to see if it increased too much overhead.
    - Also the `BytesN` custom types family

TODOs:
- [x] Make phase 0 CI pass
- [x] phase 1
- [x] Clean up
- [x] Benchmark

### CODE REVIEW

File: specs/core/1_custody-game.md [hunk 1]
```diff
@@ -682,17 +686,17 @@ Run `process_challenge_deadlines(state)` immediately after `process_reveal_deadl
...
-    for challenge in state.custody_chunk_challenge_records:
-        if get_current_epoch(state) > challenge.inclusion_epoch + CUSTODY_RESPONSE_DEADLINE:
-            slash_validator(state, challenge.responder_index, challenge.challenger_index)
-            records = state.custody_chunk_challenge_records
-            records[records.index(challenge)] = CustodyChunkChallengeRecord()
-
-    for challenge in state.custody_bit_challenge_records:
-        if get_current_epoch(state) > challenge.inclusion_epoch + CUSTODY_RESPONSE_DEADLINE:
-            slash_validator(state, challenge.responder_index, challenge.challenger_index)
+    for custody_chunk_challenge in state.custody_chunk_challenge_records:
```
Comment (hwwhww) [inline specs/core/1_custody-game.md]: Workaround for `mypy` issue of redefining variable in the different scopes (https://github.com/python/mypy/issues/5750).

**Comment by hwwhww** [comment]
Lazy benchmarking on running `make test`:
Before this PR: 106 passed in 145.90 seconds
With this PR: 106 passed in 162.60 seconds 😬

**Comment by hwwhww** [comment]
Re: benchmark
Run `make ./eth2.0-spec-tests/tests/sanity` for benchmarking
Before this PR: 106 passed in 15m 38s
With this PR: took 29m 1s  😨
/cc @protolambda

File: specs/core/0_beacon-chain.md [hunk 2]
```diff
@@ -1140,15 +1140,15 @@ def slash_validator(state: BeaconState,
     current_epoch = get_current_epoch(state)
     initiate_validator_exit(state, slashed_index)
     state.validator_registry[slashed_index].slashed = True
-    state.validator_registry[slashed_index].withdrawable_epoch = current_epoch + LATEST_SLASHED_EXIT_LENGTH
+    state.validator_registry[slashed_index].withdrawable_epoch = Epoch(current_epoch + LATEST_SLASHED_EXIT_LENGTH)
```
Comment (JustinDrake) [inline specs/core/0_beacon-chain.md]: If `current_epoch` and `LATEST_SLASHED_EXIT_LENGTH` are both of type `Epoch` do we need to wrap the sum with `Epoch( ... )`?

Comment (hwwhww) [reply specs/core/0_beacon-chain.md]: `Epoch` is defined as "EpochNumber/EpochID" and `LATEST_SLASHED_EXIT_LENGTH` is defined as "diff between Epochs".

I'm slightly in favor of the status quo but I'm open to it; for `Epoch` and `Slot`, we can change it to the more flexible definition if we want. :)

**Comment by hwwhww** [comment]
I'm not so happy with the benchmark results but it's ready to get the first review.

There are some other `uint64` in the specs, IMO we can also create custom types for some of them.
- phase 0: `genesis_time` -> `Timestamp`
- phase 1:
    - `chunk_index` -> `ChunkIndex`
    - `challenge_index` -> `ChallengeIndex`

/cc @JustinDrake what would you say?

File: specs/core/1_shard-data-chains.md [hunk 3]
```diff
@@ -79,39 +79,39 @@ class ShardBlockBody(Container):
...
 ### `ShardBlockHeader`
 
 ```python
 class ShardBlockHeader(Container):
-    slot: uint64
-    shard: uint64
+    slot: Slot
+    shard: Shard
     beacon_chain_root: Bytes32
     parent_root: Bytes32
```
Comment (JustinDrake) [inline specs/core/1_shard-data-chains.md]: All the hashes can be of type `Hash` (see PR1156 where I introduce the type).

**Comment by hwwhww** [comment]
@JustinDrake  up-to-date now with #1156

File: scripts/build_spec.py [hunk 4]
```diff
@@ -120,32 +117,46 @@ def apply_constants_preset(preset: Dict[str, Any]):
...
-        '\n' + '\n'.join(['''%s = NewType('%s', %s)''' % (key, key, value) for key, value in new_types.items()])
+    new_type_definitions = (
+        '\n\n'.join(
+            [
+                f"class {key}({value}):\n"
+                f"    def __init__(self, _x: {value}) -> None:\n"
+                f"        ...\n"
+                if value.startswith("uint")
+                # else ""
+                # else f"{key} = {value}\n"
```
Comment (protolambda) [inline scripts/build_spec.py]: old comments? correct to remove the commented lines?

File: test_libs/pyspec/eth2spec/utils/ssz/ssz_typing.py [hunk 5]
```diff
@@ -250,7 +255,7 @@ def __init__(self, *args: Iterable):
         # cannot check non-type objects, or parametrized types
         if isinstance(cls.elem_type, type) and not hasattr(cls.elem_type, '__args__'):
             for i, item in enumerate(self.items):
-                if not issubclass(type(item), cls.elem_type):
+                if not issubclass(cls.elem_type, type(item)):
```
Comment (protolambda) [inline test_libs/pyspec/eth2spec/utils/ssz/ssz_typing.py]: I think this should be `isintance(item, cls.elem_type)`. But is going to be more strict, less-newtype-workaround, and general with 1180 anyway.

File: test_libs/pyspec/eth2spec/utils/ssz/ssz_typing.py [hunk 6]
```diff
@@ -393,11 +398,27 @@ def hash_tree_root(self):
...
+    length = 48
+
+
+class Bytes96(BytesN):
+    length = 96
+
+
 # SSZ Defaults
 # -----------------------------
 def get_zero_value(typ):
```
Comment (protolambda) [inline test_libs/pyspec/eth2spec/utils/ssz/ssz_typing.py]: we'll get field/elem-type based coercion in 1180, so `get_zero_value` / `infer_type` doesn't have to be guessing types anymore. yay.

File: test_libs/pyspec/eth2spec/utils/ssz/ssz_typing.py [hunk 7]
```diff
@@ -47,8 +47,13 @@ def __new__(cls, value, *args, **kwargs):
         return super().__new__(cls, value)
 
 
-# We simply default to uint64. But do give it a name, for readability
-uint64 = NewType('uint64', int)
+class uint64(uint):
```
Comment (protolambda) [inline test_libs/pyspec/eth2spec/utils/ssz/ssz_typing.py]: de-duplicated the bounds checking in 1180 :+1:

File: test_libs/pyspec/eth2spec/utils/ssz/ssz_typing.py [hunk 8]
```diff
@@ -47,8 +47,13 @@ def __new__(cls, value, *args, **kwargs):
...
+    byte_len = 8
+
+    def __new__(cls, value, *args, **kwargs):
+        if value.bit_length() > 64:
```
Comment (djrtwo) [inline test_libs/pyspec/eth2spec/utils/ssz/ssz_typing.py]: Should we check if value less than 0 for underflows?

Comment (protolambda) [reply test_libs/pyspec/eth2spec/utils/ssz/ssz_typing.py]: doing that in new 1180 :+1:

File: scripts/build_spec.py [hunk 9]
```diff
@@ -77,36 +75,35 @@ def get_ssz_type_by_name(name: str) -> Container:
...
+                      seed: Hash,
+                      index: int,
+                      count: int) -> List[ValidatorIndex]:
     param_hash = (hash_tree_root(indices), seed, index, count)
 
-    if param_hash in committee_cache:
-        return committee_cache[param_hash]
-    else:
+    if param_hash not in committee_cache:
         ret = _compute_committee(indices, seed, index, count)
```
[the remark this answers is not in the document]
Comment (protolambda) [reply scripts/build_spec.py]: resolved in f4de5e3

File: scripts/build_spec.py [hunk 10]
```diff
@@ -77,36 +75,35 @@ def get_ssz_type_by_name(name: str) -> Container:
...
-def hash(x):
-    if x in hash_cache:
-        return hash_cache[x]
-    else:
+def hash(x: bytes) -> Hash:
+    if x not in hash_cache:
         ret = _hash(x)
-        hash_cache[x] = ret
-        return ret
+        hash_cache[x] = Hash(ret)
```
[the remark this answers is not in the document]
Comment (protolambda) [reply scripts/build_spec.py]: resolved in 207f632

File: specs/core/0_beacon-chain.md [hunk 11]
```diff
@@ -689,10 +705,10 @@ def get_shard_delta(state: BeaconState, epoch: Epoch) -> int:
 def get_epoch_start_shard(state: BeaconState, epoch: Epoch) -> Shard:
     assert epoch <= get_current_epoch(state) + 1
     check_epoch = get_current_epoch(state) + 1
-    shard = (state.start_shard + get_shard_delta(state, get_current_epoch(state))) % SHARD_COUNT
+    shard = Shard((state.start_shard + get_shard_delta(state, get_current_epoch(state))) % SHARD_COUNT)
     while check_epoch > epoch:
         check_epoch -= 1
-        shard = (shard + SHARD_COUNT - get_shard_delta(state, check_epoch)) % SHARD_COUNT
+        shard = Shard((shard + SHARD_COUNT - get_shard_delta(state, Epoch(check_epoch))) % SHARD_COUNT)
```
[the remark this answers is not in the document]
Comment (protolambda) [reply specs/core/0_beacon-chain.md]: Done in 346e61d
Will make `-`, `+`, `%` and `/` work with `uintX` `int` combinations, but not `uintX` `uintY` in #1180

File: specs/core/0_beacon-chain.md [hunk 12]
```diff
@@ -821,16 +837,19 @@ def get_shuffled_index(index: ValidatorIndex, index_count: int, seed: Hash) -> V
...
-        pivot = bytes_to_int(hash(seed + int_to_bytes(round, length=1))[0:8]) % index_count
-        flip = (pivot + index_count - index) % index_count
+    for current_round in range(SHUFFLE_ROUND_COUNT):
+        pivot = bytes_to_int(hash(seed + int_to_bytes(current_round, length=1))[0:8]) % index_count
+        flip = ValidatorIndex((pivot + index_count - index) % index_count)
         position = max(index, flip)
-        source = hash(seed + int_to_bytes(round, length=1) + int_to_bytes(position // 256, length=4))
+        source = hash(
+            seed + int_to_bytes(current_round, length=1) +
+            int_to_bytes(ValidatorIndex(position // 256), length=4)
```
[the remark this answers is not in the document]
Comment (protolambda) [reply specs/core/0_beacon-chain.md]: see 0600419
Linter is fine (locally at least, waiting for CI)

**Pull request merged**